Add a new page to present the metadata API

CodingDive opened the pull request:
Fixes https://github.com/serlo/private-issues-sso-metadata-wallet/issues/26


preview: https://frontend-git-metadata-api-presentation-serlo.vercel.app/metadata

There are still some todos left

- [x] Fix types of GraphQl examples so that they can be copy-pasted without errors and put them on the Wiki
- [x] Decide what goes into the [Wiki](https://github.com/serlo/documentation/wiki/Metadata-API) and which information should live on the webpage (e.g properties, AMB standard, code examples, etc.)
- [x] Iterate on UI and content with Simon and Gregor 
- [x] Call to Action (x2) 
- [x] Finish the Wiki page so that it has content-parity with https://nbviewer.org/github/serlo/lenabi/blob/20b946ff9f1205444f256995dfd776fd203b6c3c/src/Prototype%20of%20metadata%20API%20for%20serlo.org%20%28LENABI%29.ipynb#format

Commit:
Add a new page to present the metadata API

Commit:
Center tabs

Commit:
Remove old id & trigger rebuild

hugotiburtino commented:
Since it is GraphQL, the examples are a bit more complex.
See https://nbviewer.org/github/serlo/lenabi/blob/20b946ff9f1205444f256995dfd776fd203b6c3c/src/Prototype%20of%20metadata%20API%20for%20serlo.org%20%28LENABI%29.ipynb for working examples

Here an example that would work
```python
import requests

payload = {
    "first": 10,
}

response = requests.post(
    "https://api.serlo.org/graphql",
    headers={"Content-Type": "application/json"},
    json={
        "query": """
        query($first: Int) {
            metadata {
                entities(first: $first) {
                    nodes
                }
            }
        }
    """,
        "variables": {
            "first": payload["first"],
        },
    },
)

print(response.json())
```
---

Idea: https://github.com/orgs/serlo/projects/35?pane=issue&itemId=28489612

 ---

But aside from that: I think the more technical documentation should go anyway to another place (v.g. Github Wiki), if it is here about speaking to possible POs.

hugotiburtino commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,337 @@
+            </h2>
+            <div className="text-left mx-side">
+              <p className="mt-4 text-xl leading-cozy flex-1">
+                Join our growing community by integrating our content through
+                the Serlo Metadata API into your apps and change the future of
+                education with us!
```
I think it may confuse readers, that may think that it is like our content API.

```suggestion
                Find more easily our contents using Serlo Metadata API!
```

hugotiburtino commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,337 @@
+  )
+}
+
+const codeSnippets: Record<string, string> = {
```
As said, these code snippets would have to be changed in order to really work. 
But my suggestion is to introduce a simple 'Docs' link, which would redirect to a Github Wiki page, where devs could get a fully explanation on how to use our GraphQL query.

CodingDive replied:
Right, I forgot that the database layer is behind https://github.com/serlo/api.serlo.org :see_no_evil: 

Good idea, I'll create a Github wiki page for it.

hugotiburtino commented:
And something as contact info like
![image](https://github.com/serlo/frontend/assets/45924645/24b131d0-12de-4a21-94a1-c30e4ac488e0)
(but Kulla instead), would be good also 😊

kulla commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,337 @@
+  )
+}
+
+const exampleJsonResponse = `
```
This data can be fetched from the API via `getServerSideProps()`

kulla commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,337 @@
+
+print(response.json())
+`,
+  Rust: `
```
I would add a `curl` example as well.

kulla reviewed (changes requested):
Just a short review (will do a proper review later). So far:

* The endpoint needs to be described as in https://lenabi.serlo.org/metadata-api (description of parameters)
* section "How to update your data"
* I would describe which kind of data we return a little bit (maybe a list of the properties with a link to their definitions?)

Commit:
Implement some feedback of Kulla and Hugo

- Add contact person (Kulla)
- Add example for curl
- Change payload to GraphQL
- Attempt to make the distinction to the content API clear
- Link to wiki for more technical documentation

Commit:
Implement first iteration of feedback from Gregor

* Remove old code
* Fix build / eslint error
* Create Call to Action that is used at the top and the bottom linking to the docs

Commit:
Fix responsiveness of bird + text on mobile

Commit:
Iterate with Gregor on copywriting and CTA

Commit:
Iterate on alignment of CTA button

Commit:
Add experimental wing animation of Botho

Commit:
Increase spacing and move code around

CodingDive commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,182 @@
+          <EditorContact lastName="Kulla" />
+        </div>
+      </div>
+      <style jsx>
+        {`
+          :global(.landing-button-with-wings):after,
+          :global(.landing-button-with-wings):before {
+            content: ' ';
+            background: url('/_assets/img/landing/wing-left.svg') no-repeat;
+            position: absolute;
+            margin-top: -0.6rem;
+            width: 4rem;
+            height: 4rem;
+            pointer-events: none;
+            opacity: 0;
+            transition: opacity ease-in 0.2s;
+          }
+          :global(.landing-button-with-wings):after {
+            margin-left: 1rem;
+            transform: scaleX(-1);
+          }
+          :global(.landing-button-with-wings):before {
+            margin-left: -5rem;
+          }
+          :global(.landing-button-with-wings):hover:after,
+          :global(.landing-button-with-wings):hover:before {
+            opacity: 1;
+          }
+          :global(.landing-button-with-wink):after,
+          :global(.landing-button-with-wink):before {
+            background: url('/_assets/img/landing/wink-left.svg') no-repeat !important;
+            margin-top: -2rem !important;
+            background-size: 65% !important;
+          }
+        `}
+      </style>
```
Pretty cool code @elbotho! Gregor gave me the idea to also add the wings here. I copied it alongside a lot of other code from the editor-presentation. Is it possible to rewrite this within pure tailwind now that we have decided to not use `styled-components` or `jsx` anymore?

elbotho replied:
see slack

elbotho replied:
Currently you need to add here to get the extrabold font style:
``` css
          @font-face {
            font-family: 'Karmilla';
            font-style: bolder;
            font-weight: 800;
            src: url('/_assets/fonts/karmilla/karmilla-bolder.woff2')
                format('woff2'),
              url('/_assets/fonts/karmilla/karmilla-bold.woff') format('woff');
            font-display: swap;
          }
```

elbotho commented on src/components/pages/editor/editor-contact.tsx:
```diff
@@ -0,0 +1,49 @@
+
+import { teamData } from './editor-team'
+
+type IEditorContactProps =
```
(code style) currently we don't prefix interfaces with `I`

elbotho commented on src/components/pages/editor/editor-contact.tsx:
```diff
@@ -0,0 +1,49 @@
+  | { firstName: string; lastName?: string }
+  | { firstName?: string; lastName: string }
+
+export const EditorContact: FC<IEditorContactProps> = ({
```
(code style) currently we always always use `function EditorContact` for components instead of `const EditorContact` and almost never use `React.FC` to type them.

elbotho commented on src/components/pages/editor/editor-contact.tsx:
```diff
@@ -0,0 +1,49 @@
+}) => {
+  if (!firstName && !lastName) {
+    throw new Error('firstName or lastName must be provided')
+  }
```
maybe move error below the `find` if no contact was found with the given names?

elbotho replied:
(or remove it, while developing ts should catch this anyway, right?)

CodingDive replied:
Both args missing should have been caught by ts indeed. As you suggested, I moved it after the find to have an extra check in case our constant `teamData` changes

elbotho commented on src/components/pages/editor/editor-team.tsx:
```diff
@@ -44,8 +44,8 @@ export const teamData: TeamDataEntry[] = [
-    extraLinkUrl: '',
-    extraLinkText: '',
+    extraLinkUrl: 'mailto:kulla@serlo.org',
+    extraLinkText: 'kulla@serlo.org',
```
This has an unwanted side effect: Kulla now also has a mail address on the `/editor` page.
Maybe you find an easy way to separate this.

CodingDive replied:
Added an extra optional property to the contacts that I only set to `true` for Wolfgang, Sebastian and Simon

elbotho commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,182 @@
+  alignment?: 'center' | 'left'
+}
+
+const CallToAction: React.FC<CallToActionProps> = ({ alignment = 'left' }) => {
```
(same code style remark as above)

elbotho committed:
add extrabold,  whitespace change

elbotho reviewed (commented):
Thanks, Code looks good!
Added some comments, most of them regarding code style or suggestion.
Take them or leave them, not critical.

elbotho committed:
refactor eyebrows and wings 🥸🦋

elbotho committed:
add extrabold karmilla to default fonts

elbotho committed:
refactor underlined

elbotho commented:
@CodingDive sorry for hijacking your PR, did some style refactoring. Upsides: less code duplication, your `styled-jsx` problem is gone and you get the extrabold titles and pretty underlined heading for free :)
Now it's all yours again.

Commit:
Implement feedback of Botho (mostly code style)

Commit:
Small iterations on content with feedback of Botho

elbotho committed:
use tw for image hack

kulla commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,150 @@
+            <div className="flex-1 mt-5">
+              <b className="text-brand font-handwritten text-4xl">20,000+</b>
+              <br />
+              Educational resources have been created with the{' '}
```
```suggestion
               X educational resources are available over the Metadata API
```

You can find X by importing the current staging data in `database-layer` with `yarn mysql:import-anonymous-data` and count the number of overall entries with `yarn fetch EntitiesMetadataQuery ...` or by executing the underlying SQL without limit

CodingDive replied:
The goal of this was to link back to the Serlo Editor.
I could add a third point saying "10.000+ educational resources are available over the Metadata API" but I'd leave this the way it is.

Here are some interesting numbers, some of which we could display too.
```
DE 10922
out of which
  applet 144
  article 2967
  course 598
  text-exercise 4088
  text-exercise-group 2691
  video 434
EN 686
ES 553
```

kulla replied:
> The goal of this was to link back to the Serlo Editor.

I don't understand this -> Why do we want to link to the Serlo editor here / Why do we want to hint people to the Serlo editor? Anyhow then this link wouldn't be understandable for me.

CodingDive replied:
I thought for better SEO and because the same people that are interested in the Metadata API might also be interested in other Serlo offerings such as our editor. Would you like me to remove the link or first discuss it with Simon on Friday?

kulla commented on src/components/pages/metadata-api/metadata-api-presentation.tsx:
```diff
@@ -0,0 +1,150 @@
+            <div className="flex-1 mt-5">
+              <b className="text-brand font-handwritten text-4xl">1 Mio+</b>
+              <br />
+              Users per month already learn with Serlo
```
```suggestion
              Users per month already use those materials
```

kulla commented on src/components/pages/editor/editor-contact.tsx:
```diff
@@ -0,0 +1,45 @@
+  | { firstName: string; lastName?: string }
+  | { firstName?: string; lastName: string }
+
+export function EditorContact({ firstName, lastName }: EditorContactProps) {
```
I would define it as

```typescript
function EditorContact(contact: TeamDataEntry) {

}
```

In the editor you can call it with the editor contacts and on metadata with my contact information. Then you do not need `isEditorContactPerson`.

Commit:
Implement most feedback of Kulla

Commit:
Implement feedback of Lama meeting

* Increase spacing
* Open link in new tab

Commit:
Vertically center text in yellow box

Commit:
Merge branch 'staging' into metadata-api-presentation

kulla committed:
refactor(editor): Add `teamDataKulla`

kulla committed:
refactor(editor): Add teamDataSimon

kulla commented on src/pages/metadata.tsx:
```diff
@@ -0,0 +1,11 @@
+import { FrontendClientBase } from '@/components/frontend-client-base'
```
Shall we have the url `serlo.org/metadata-api`?

CodingDive replied:
I find `serlo.org/metadata` sexier but I'm open to it.

CodingDive committed:
Merge pull request #2454 from serlo/refractor-metadata-api

Refractor metadata api

Commit:
Implement some feedback of Martin with catchy phrase

elbotho committed:
Merge branch 'staging' into metadata-api-presentation

elbotho committed:
run prettier after merge

Commit:
Change picture

Commit:
Implement feedback of Kulla and Kathi to list # of entities

Commit:
Merge branch 'staging' into metadata-api-presentation
